sklearn/impute.py

```diff
@@ -138,30 +203,40 @@ def fit(self, X, y=None):
         -------
         self : SimpleImputer
         """
-        # Check parameters
-        allowed_strategies = ["mean", "median", "most_frequent"]
-        if self.strategy not in allowed_strategies:
-            raise ValueError("Can only use these strategies: {0} "
-                             " got strategy={1}".format(allowed_strategies,
-                                                        self.strategy))
+        X = self._validate_input(X)
 
-        X = check_array(X, accept_sparse='csc', dtype=FLOAT_DTYPES,
-                        force_all_finite='allow-nan'
-                        if self.missing_values == 'NaN'
-                        or np.isnan(self.missing_values) else True)
+        # default fill_value is 0 for numerical input and "missing_value"
+        # otherwise
+        if self.fill_value is None:
+            if X.dtype.kind in ("i", "u", "f"):
+                fill_value = 0
+            else:
+                fill_value = "missing_value"
+        else:
+            fill_value = self.fill_value
+
+        # fill_value should be numerical in case of numerical input
+        if (self.strategy == "constant" and
+                X.dtype.kind in ("i", "u", "f") and
+                not isinstance(fill_value, numbers.Real)):
+            raise TypeError("'fill_value'={0} is invalid. Expected a numerical"
+                            " value when imputing numerical"
+                            " data".format(fill_value))
```

Comment:
As we switched to `ValueError` above we should probably be consistent and do it here as well.

I think value error is fine because this is already what is raised by most models who use the standard `check_X_y` validation:

```python
>>> from sklearn.linear_model import LogisticRegression
>>> LogisticRegression().fit([['invalid', 'invalid']], [0, 1])
Traceback (most recent call last):
  File "<ipython-input-6-91496e8d832f>", line 1, in <module>
    LogisticRegression().fit([['invalid', 'invalid']], [0, 1])
  File "/home/ogrisel/code/scikit-learn/sklearn/linear_model/logistic.py", line 1217, in fit
    order="C")
  File "/home/ogrisel/code/scikit-learn/sklearn/utils/validation.py", line 671, in check_X_y
    ensure_min_features, warn_on_dtype, estimator)
  File "/home/ogrisel/code/scikit-learn/sklearn/utils/validation.py", line 494, in check_array
    array = np.asarray(array, dtype=dtype, order=order)
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/numpy/core/numeric.py", line 492, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: could not convert string to float: 'invalid'
```